### ISSUE TITLE
Kerbalism Core

### ISSUE THREAD

**Issue by SirMortimer** [Kerbalism/Kerbalism#426]
#424 and a couple of issues related to the new science mode got me thinking.

**Note that this is in no way thought through or a finished design proposal, it's just brainstorming that I put here because it's more permanent / easier to access than discord:**
 
## Instead of adding more and more features to Kerbalism, split it into smaller mods.

Namely: 
- Core: a mod that provides just barebone functionality for background processing and a user interafce. Responsible to handle resource processing (EC, food, oxygen, you name it) in a reliable way that works well with high time warp. Also handles data and data transmission. The core mod could do to background processing what Community Resource Pack does to resources: unify the system across mods, so that those mods wouldn't have to re-implement background processing all the time.

- Sensors: provide a way to get reliable and good approximations of vessel parameters (current and for some time into the past, which would be used for time warp) that we need for science: temperature, radiation, sun angle, altitude, stuff like that.
- Comms: provide a way to feed communication parameters to be used into the Core. We already have that and use it internally for RT, Real Antennas is the first external mod to use it.

We then could move some of the functionality that currently is within Kerbalism into those other mods, which would make those mods smaller (easier to maintain, bugfix, update) or exchangeable with other mods (use Kerbal Health, Test Flight or Snacks instead of the Kerbalism equivalents):
- Life support (eating, breathing, processes) goes into "Kerbalism LSS".
- Reliability goes into "Kerbalism Reliability" (or a better, fancier name)
- Radiation model + Sun CME sims could go into a new mod "Kerbal Space Weather"

![cbp](https://user-images.githubusercontent.com/43166475/59330814-ca913900-8cf2-11e9-84df-3b7f7bfdee92.png)

Note: this is all just brainstorming and in no way accurate or a well thought out architectural design. just an idea to put it out there. I know that taking Kerbalism apart would be a LOT of work, and will be a while until we have it back at it's current state. Might be a long while.
Not sure if it is woth the effort.

**But: we can't sustain the current rate of feature creep.** The code base gets too big for effective maintenance, especially for a hobby project. Adding too much will kill it over time.

**Comment by UsernamesAreAlwaysTaken** [comment]
I agree.
This would be taking the USI approach. With all its flaws, RoverDude had the right idea: Break it down into smaller independent bits, allowing users to install only what they want for one, and two: should make maintenance more straightforward.

We're past Kerbalism being a life support mod only, and the code base getting larger and larger, we could get people working only on a specific bit (if we get more people willing to contribute). This way, If an issue pops up, whoever is responsible for that specific part would be more likely to know exactly what's happening and be able to fix it.

Take science configs for example. I designed that clusterfuck, and i know exactly what and where it is happening, so it takes me very little time to pinpoint the problem (think basevalue = sciencecap issue, i was able to fix the configs in a couple of minutes once i knew what was going on).

In short, i fully support this proposal, with the caveat that certain bits are assigned to certain people, so they can learn and understand properly what's going on in the code for ease of dealing with issues and bugs.
Finding more people to contribute though...

**Comment by gotmachine** [comment]
I also think some Kerbalism features should be splitted as "addons", especially those that aren't directly related to the core background processing / life support / resource processes. What I consider core features are :
1. The processes / rules config framework, and associated code
2. The background processing framework : the resource simulator, vessel environment simulator, and partmodule processing.
3. The UI framework, the planner and vessel status panel
4. Life support / habitat / resource related rule modifiers and partmodules
5. The comms tweaks built on top of commnet

Code for 1. is quite stable and mature. 2. has several core issues, and need a complete rewrite to achieve a stable and maintainable framework, that is extensible by other plugins. 3. is crippled by performance issues due to the use of the old OnGui/GUILayout system. 4 has some minor bugs but is overall working well, but maintainability would benefit from the 2. rewrite. I'm not familiar with the code for 5., so I won't comment.

I don't think there is much point in splitting the life support away. Kerbalism is superior to every other LS plugin, and even a LS plugin that is very different like Snacks can be replicated 1:1 in a few hours using a custom profile. Having it as core allow us to take many shortcuts instead of developing complex interfaces which in my opinion nobody will ever use.
For the same reason, I think Stress should stay as core

Features that are definitely good candidate to get splitted are reliability, radiation and science.

To expand on the issues on point 2. : 
- The core resource background processing system is working well, but it currently is crippled by unadequate handling of pure consumers :
    - Non-recipe consumers are currently implemented by checking if the resource amount was > 0 at last update and consuming a fixed amount, causing unstability in the resource rates, causing inconsistencies in recipes/rules/processes execution.
    - This make any non-resource output (a few ex : transmission, science data, active shielding, gravity ring...) inconsistent (and sometimes plain wrong), as there is no way to scale the output against the available resources.
    - I've not investigated much, but it seems the only way to handle this is by implementing "virtual resources" in recipes, and forbid any resource consumption outside the recipes. For example, experiment "X" would declare a recipe that has "EC" as input and "Xdata" as output. There is already something like that done for "cures", but we need to generalize it to everything. The big issue is that to use this, we need to split the module updates into a "pre-ressource" update (where for example, experiment-specific checks are performed, and the recipe registered) and a "post-ressource" update that happen after all recipes/processes/rules have been executed. This mean refactoring how we do background processing for partmodules, see next point.
- The current way of handling background processing for partmodules is a mess. 
  - First, the code duplication and fragmentation. Functionality is splitted in 5 places : fixedUpdate, backgroundUpdate, various hacks in the Background class, the sensors for the Kerbalism UI, and the infamous planner mess. 
  - Second, the performance impact of manipulating the serialized protomodule data (instead of real alive objects) is becoming quite important as more and more modules are implemented. 

What we need is our own partmodule framework that work "on top" of the stock partmodules, implemented as a base abstract class and interfaces. These modules would be kept alive regardless of the loaded/unloaded state, and all the logic and data would be stripped from the stock partmodules and put in them. That would allow implementing the mentioned above pre-ressource/post-ressource update methods, and standard methods like load/save, interfaces for things like module info, planner info, sensor info, etc. It should be possible to make that flexible enough to communicate with stock/other plugins partmodules when we want to "hack" them without doing a complete rewrite.

- The vessel cache updates are currently desynchronized with the vessel updates. The only reason for this is due to the history of the mod, it has no justification and it need to be rewritten in favor of an update-synchronized "vessel data" object. 
    - The first reason is related to the above : we need our own "vessel" class, independent from the stock loaded/unloaded vessel/protovessel, to hold our own "part" objects and our own "module" objects. We need to abstract the stock objects to fit our own needs, to centralize the persistence, to maintain stable and easy accessed references to our own objects, and to the related stock objects/classes. This will be a lot cleaner and efficient than the dozens of various queries we are doing everywhere to the stock objects using a mix of references, string IDs and dictionaries, anonymous caches, etc.
    - Second reason is that we need a better vessel environment evaluation at high time-warp speeds. It's in my opinion the most problematic issue currently, as this affects many aspects of the simulation, the most visible being the solar panel EC output inconsistencies. I'm currently doing experiementations with a sub-update simulation at fixed timesteps. It's a brute force approach : I'm evaluating the vessel environnement every 60 seconds in the past along the orbit, regardless of the timewarp speed. I don't think there is any other realistic solution (the complexity of doing analytical estimations for everything is way beyond our skills, available dev time and computational power), and it has the potential to work for everything and can potentially be optimized by multithreading.

**Comment by SirMortimer** [comment]
I think with KSP2 on the horizon this discussion is pretty much moot at this point. It's definitely something to consider for a KSP2-Kerablism (if there will be one), but actively splitting the existing mod for the existing KSP... not going to happen. It's a huge task, effort better spent elsewhere.
